plugins/builds/index.js

```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
+                jobFactory,
+                buildFactory,
+                eventFactory,
+                pipelineId,
+                jobName: nextJobName,
+                username,
+                scmContext,
+                build, // this is the parentBuild for the next build
+                baseBranch: event.baseBranch || null,
+                parentBuilds
+            };
+            const externalBuildConfig = {
+                pipelineFactory,
+                eventFactory,
+                externalPipelineId,
+                externalJobName,
+                parentBuildId: build.id,
+                parentBuilds,
+                causeMessage: `Triggered by sd@${pipelineId}:${currentJobName}`
+            };
+
+            /* CREATE AND START NEXT BUILD IF ALL INTO 3 SCENARIOS
+             * 1. No join
+             * 2. ([~D,B,C]->A) currentJob=D, nextJob=A, joinList(A)=[B,C]
+             *    joinList doesn't include D, so start A
+             * 3. ([~D,B,C]-> sd@123:A) currentJob=D, nextJob=sd@123:A, joinList(A)=[sd@111:B,sd@111:C]
+             *    joinList doesn't include sd@111:D, so start A
+             */
+            // current job can be "external" in nextJob's perspective
+            const currentJobNotInJoinList = !joinListNames.includes(currentJobName) &&
+                !joinListNames.includes(`sd@${pipelineId}:${currentJobName}`);
+
+            if (joinList.length === 0 || currentJobNotInJoinList) {
+                if (!isExternal) {
+                    return createInternalBuild(internalBuildConfig);
+                }
+
+                return createExternalBuild(externalBuildConfig);
+            }
+
+            let nextBuild;
+
+            /* CHECK WHETHER NEXT BUILD EXISTS */
+            // if next build is external, check the latest build
+            if (isExternal) {
+                const p = await pipelineFactory.get(externalPipelineId);
+                const jobArray = await p.getJobs({ params: { name: externalJobName } });
+                const j = await jobFactory.get(jobArray[0].id);
+
+                nextBuild = await j.getLatestBuild({ status: 'CREATED' });
+
+            // if next build is internal, look at the finished builds for this event
+            } else {
+                let finishedInternalBuilds;
 
-                // Just start the build if falls in to these 2 scenarios
-                // 1. No join
-                // 2. ([~D,B,C]->A) currentJob=D, nextJob=A, joinList(A)=[B,C]
-                //    joinList doesn't include C, so start A
-                if (joinList.length === 0 || !joinListNames.includes(currentJobName)) {
-                    return createBuild(buildConfig);
+                if (!event.parentEventId) {
+                    finishedInternalBuilds = await event.getBuilds();
+
+                // If parent event id, merge parent build status data and
+                // rerun all builds in the path of the startFrom
+                } else {
```

Comment:
Similar to old logic here: https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/builds/index.js#L343-L365